pkg/geo/geogfn/topology_operations_test.go

```diff
@@ -0,0 +1,152 @@
+// licenses/APL.txt.
+
+package geogfn
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/stretchr/testify/require"
+	"github.com/twpayne/go-geom"
+)
+
+func TestCentroid(t *testing.T) {
+	testCases := []struct {
+		wkt                      string
+		expectedSphereCentroid   string
+		expectedSpheroidCentroid string
+	}{
+		{
+			wkt:                      `POINT (1.0 1.0)`,
+			expectedSphereCentroid:   `POINT (1.0 1.0)`,
+			expectedSpheroidCentroid: `POINT (1.0 1.0)`,
+		},
+		{
+			wkt:                      `SRID=4326;POINT (1.0 1.0)`,
+			expectedSphereCentroid:   `SRID=4326;POINT (1.0 1.0)`,
+			expectedSpheroidCentroid: `SRID=4326;POINT (1.0 1.0)`,
+		},
+		{
+			wkt:                      `LINESTRING (1.0 1.0, 2.0 2.0, 3.0 3.0)`,
+			expectedSphereCentroid:   `POINT (1.99961910254054 2.00007602871312)`,
+			expectedSpheroidCentroid: `POINT (1.99962062254316 2.00007754894762)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0))`,
+			expectedSphereCentroid:   `POINT (0.499999916440062 0.500006427946943)`,
+			expectedSpheroidCentroid: `POINT (0.500000001396299 0.500006342987471)`,
+		},
+		{
+			wkt:                      `POLYGON ((1.0 1.0, 0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0))`,
+			expectedSphereCentroid:   `POINT (0.499999916440062 0.500006427946943)`,
+			expectedSpheroidCentroid: `POINT (0.500000001396299 0.500006342987471)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 80.0 0.0, 80.0 80.0, 0.0 0.0))`,
+			expectedSphereCentroid:   `POINT (43.8350223319301 30.5453774409703)`,
+			expectedSpheroidCentroid: `POINT (43.8350223319301 30.5453774409703)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0, 0.0 0.0), (0.1 0.1, 0.9 0.1, 0.9 0.9, 0.1 0.9, 0.1 0.1))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `POLYGON ((1.0 1.0, 0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0), (0.1 0.1, 0.9 0.1, 0.9 0.9, 0.1 0.9, 0.1 0.1))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `POLYGON ((0.0 1.0, 0.0 0.0, 1.0 0.0, 1.0 1.0, 0.0 1.0), (0.9 0.9, 0.1 0.9, 0.1 0.1, 0.9 0.1, 0.9 0.9))`,
+			expectedSphereCentroid:   `POINT (0.500000001981181 0.500010402851408)`,
+			expectedSpheroidCentroid: `POINT (0.499999862652707 0.500010542184438)`,
+		},
+		{
+			wkt:                      `MULTIPOINT (0.0 0.0, 1.0 10.0, 2.0 30.0)`,
+			expectedSphereCentroid:   `POINT (0.953002819213836 13.2963208764279)`,
+			expectedSpheroidCentroid: `POINT (0.953002819213836 13.2963208764279)`,
+		},
+		{
+			wkt:                      `MULTILINESTRING ((1.0 1.0, 2.0 2.0, 3.0 3.0), (0.0 0.0, 1.0 10.0, 2.0 30.0))`,
```

Comment:
mind adding a few negative coordinates for tests?

Reply:
Sure